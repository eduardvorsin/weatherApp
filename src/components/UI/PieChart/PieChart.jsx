import PropTypes from 'prop-types';
import classes from './PieChart.module.css';

export default function PieChart({
  value,
  className,
  testId,
}) {
  return (
    <div
      className={`${classes['pie-chart']} ${className}`}
      data-testid={testId}
    >
      <svg
        className={classes['pie-chart__icon']}
        viewBox="0 0 512 512"
        fill="none"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          className={classes['pie-chart__circle']}
          cx="255"
          cy="255"
          r="245.25"
        />
        <circle
          className={classes['pie-chart__circle']}
          style={{ '--value': `${value}` }}
          cx="255"
          cy="255"
          r="245.25"
        />
      </svg>
      <p className={classes['pie-chart__text']}>
        {value}
        %
      </p>
    </div>
  );
}

PieChart.defaultProps = {
  className: '',
  value: 75,
  testId: '',
};

PieChart.propTypes = {
  className: PropTypes.string,
  value: PropTypes.number,
  testId: PropTypes.string,
};
