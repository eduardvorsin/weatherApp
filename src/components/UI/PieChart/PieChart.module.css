.pie-chart {
  position: relative;
  overflow: hidden;
}

.pie-chart__circle {
  --pie-chart-dasharray: 1545;
  --pie-chart-dashoffset: 1545;

  stroke: #ffffff;
  stroke-width: 10;
  stroke-dasharray: var(--pie-chart-dasharray);
  stroke-dashoffset: var(--pie-chart-dashoffset);
}

.pie-chart__circle:first-child {
  stroke-dashoffset: 0;
  stroke: rgb(55, 55, 55);
}

.pie-chart__circle:last-child {
  stroke-dashoffset: calc(var(--pie-chart-dasharray) - (var(--pie-chart-dashoffset) * var(--value))/100);
  animation: animate-circle 2s cubic-bezier(0, 0.23, 1, .1);
}

.pie-chart__text {
  margin: 0;

  padding: 15px;
  text-align: center;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

@keyframes animate-circle {
  from {
    stroke-dashoffset: var(--pie-chart-dashoffset);
  }
}
