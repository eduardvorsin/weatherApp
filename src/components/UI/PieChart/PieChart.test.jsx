import { render, screen } from '@testing-library/react';
import PieChart from './PieChart';

describe('PieChart component tests', () => {
  it('renders correctly', () => {
    render(<PieChart
      value={50}
      className="test-chart"
    />);

    expect(screen.getByText('50%')).toBeInTheDocument();
  });

  it('renders correctly without props', () => {
    render(<PieChart />);

    expect(screen.getByText('75%')).toBeInTheDocument();
  });

  it('the classname is correctly assigned to the element', () => {
    render(<PieChart
      className="test-chart"
      testId="pie-chart"
    />);

    expect(screen.getByTestId('pie-chart')).toHaveClass('test-chart');
  });

  it('snapshot with default value', () => {
    render(<PieChart
      testId="pie-chart"
    />);

    expect(screen.getByTestId('pie-chart')).toMatchSnapshot();
  });

  it('snapshot with mock value', () => {
    render(<PieChart
      value={33}
      testId="pie-chart"
    />);

    expect(screen.getByTestId('pie-chart')).toMatchSnapshot();
  });
});
