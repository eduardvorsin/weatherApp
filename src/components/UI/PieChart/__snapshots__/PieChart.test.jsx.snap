// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart component tests snapshot with default value 1`] = `
<div
  class="pie-chart "
  data-testid="pie-chart"
>
  <svg
    class="pie-chart__icon"
    fill="none"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="pie-chart__circle"
      cx="255"
      cy="255"
      r="245.25"
    />
    <circle
      class="pie-chart__circle"
      cx="255"
      cy="255"
      r="245.25"
      style="--value: 75;"
    />
  </svg>
  <p
    class="pie-chart__text"
  >
    75
    %
  </p>
</div>
`;

exports[`PieChart component tests snapshot with mock value 1`] = `
<div
  class="pie-chart "
  data-testid="pie-chart"
>
  <svg
    class="pie-chart__icon"
    fill="none"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="pie-chart__circle"
      cx="255"
      cy="255"
      r="245.25"
    />
    <circle
      class="pie-chart__circle"
      cx="255"
      cy="255"
      r="245.25"
      style="--value: 33;"
    />
  </svg>
  <p
    class="pie-chart__text"
  >
    33
    %
  </p>
</div>
`;
